Mom will initiate the dialogue sequence with server

nithinj opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Describe Bug or Feature
Overview

There is a sequence of dialogue taking place between the server and mom when they come up. This change focuses on reversing the initial exchange to reduce the overhead on the server and making it adaptable for future scalability enhancements.

![image](https://user-images.githubusercontent.com/16252556/78898028-1c087280-7a41-11ea-8a6d-0c78eb6de5cd.png)
![image](https://user-images.githubusercontent.com/16252556/78898033-1f036300-7a41-11ea-934d-7b4c45160c8d.png)


    Initial Dialogue Exchange:

Initial dialogues exchange is a number of message exchange happens between server and mom before the server shows the node as up. The number of message exchange is reduced. Following is a sequence diagram of how mom-server interaction appeared before and now.

    IS_HELLOSVR

When the server comes up it sends IS_HELLO to all the moms in the cluster. Mom will reply back with a list of jobs and state information. This is not an efficient method as the server has to deal diligently to keep track of the moms who have not responded yet and continue sending IS_HELLO in a repeated interval. The same can be achieved by sending the first hello message from MOM without incurring overhead on the server. Mom will be initiating the dialogue with IS_HELLOSVR. The server will then check whether the receivers network resolved hostname is available in its database corresponding to a mom.

    IS_REPLYHELLO

Server will send IS_REPLYHELLO in response to IS_HELLOSVR from mom. This contains rpp values as well as the IP address of all moms for single-server mode which was being sent as part of IS_NULL and IS_CLUSTER_ADDRS.

    IS_REGISTERMOM

Mom will send IS_REGISTERMOM in response to IS_REPLYHELLO. It will contain job and vnode details. This will serve the purpose of IS_HELLO4, IS_UPDATE2 and IS_MOM_READY.

    IS_NULL

IS_NULL is a repeated heartbeat from the server to mom with rpp values. The message body contains rpp_retry and rpp_highwater.

Whenever rpp values are updated in the server it needs to update all the moms with the new set of values. But periodic ping will be removed as there are no use cases it can solve.

    IS_CLUSTERADDRS

IS_CLUSTERADDRS will send a list of addresses from server to mom which can be used for inter mom communication (More info on IS_CLUSTERADDRS). This will not be enabled for multi-server mode. Users have to use supported alternative security models as reserved ports based authentication is not secure for inter-mom communication.

    Frequency of HELLOSVR

Mom will start with quick bursts of IS_HELLOSVR followed by slower intervals. This strategy will make sure a faster connection if the server is already up and does not blow up the network with tons of requests. Mom will attempt to connect every 64 seconds in the worst case.

    Failover:

Mom will be treating two of the servers in its list with equal precedence and will be sending hello randomly to one of them. It can receive a reply back only from the active server. Mom will follow the same in the future when we have active-active failover.

    Server talking to an alien Mom:

When a server gets a job request from the scheduler which is supposed to run on a mom it does not know yet (thus alien mom), it will look into the database and find the required information. It then opens a connection stream with that mom and sends the job.

    The server attempts to talk to the mom before mom attempts to connect

This can happen due to operations such as delete_job or run_job reaching at the server before mom has reached out. Server will then attempt to open a connection via tpp and send the request to mom. Upon receiving the request mom will initiate hello exchange as well as initiating hello exchange sequence. This will succeed only if the mom is up and connected to comm.


#### Link to Design Doc
[project documentation area](https://pbspro.atlassian.net/wiki/spaces/PD/pages/1457029123/MOM+will+initiate+the+dialogue+sequence+with+server.)


#### Attach Test and Valgrind Logs/Output
<!--- Please attach your test log output from running the test you added (or from existing tests that cover your changes) -->
<!--- Don't forget to run Valgrind if appropriate and attach the resulting logs -->
<img width="1676" alt="image" src="https://user-images.githubusercontent.com/16252556/77554290-44626f80-6e8c-11ea-8a8b-ef5f6ecf1a0b.png">




<!--- Pull Request Guidelines: [Pull Request Guidelines](https://pbspro.atlassian.net/wiki/spaces/DG/pages/1187348483/Pull+Request+Guidelines) -->

agrawalravi90 commented on src/resmom/catch_child.c:
```diff
@@ -2063,16 +2017,12 @@ send_restart(void)
+	else
+		svr = get_servername(&port);
+	
+	send_hellosvr_rpp(svr, port);
```
should this be send_hellosvr_tcp? I think _tcp functions use TPP, we should keep function names consistent with the rest of the code.

agrawalravi90 commented on src/resmom/mom_main.c:
```diff
@@ -8194,6 +8194,46 @@ net_down_handler(void *data)
+ * @param[in] mode -	reset mode is to bring it back to bursting mode
+ * 
+ * @return int
+ * @retval >0 : time mom should wait before sending the next hello
```
is this time in seconds? Would be good to mention the unit here

nithinj replied:
done!

agrawalravi90 commented on src/resmom/mom_main.c:
```diff
@@ -8194,6 +8194,46 @@ net_down_handler(void *data)
+	}
+
+	if (cnt == 0) {
+		if (delta & ~max_delta_mask) {
```
If i understand this correctly, the max value of delta is 256 seconds. Can you please mention that somewhere as a comment? I think you should also mention this in the design.

nithinj replied:
64s. I've added a comment also updated in the design doc.

agrawalravi90 commented on src/resmom/mom_main.c:
```diff
@@ -9738,9 +9779,16 @@ main(int argc, char *argv[])
+		time_now = time(NULL);
+		if (server_stream == -1) {
+			if (time_now > time_last_hello) {
+				time_last_hello = time_now + time_delta(MOM_DELTA_NORMAL);
```
It should be called "time_next_hello" right? It's the time that you'll send hello next.

nithinj replied:
make sense!

subhasisb commented on src/include/libpbs.h:
```diff
@@ -300,7 +300,7 @@ struct batch_reply {
 #define PBS_BATCH_CopyHookFile	85
 #define PBS_BATCH_DelHookFile	86
-#define PBS_BATCH_MomRestart	87
+/* Unused -- #define PBS_BATCH_MomRestart	87 */
```
> Why not put these in an enum so that we can remove obsolete symbols instead of commenting them out?

We cannot put them in an enum, since if we delete an entry the numbers will change and backward compatibility will be compromised.

agrawalravi90 replied:
Oh, are these numbers externally visible, stable interfaces?

subhasisb commented on src/include/libutil.h:
```diff
@@ -323,3 +323,5 @@ perf_stat_stop(char *instance);
 #endif
 #endif
+
+int get_max_servers();
```
I believe Brem's sharding code is going to add these functions?

nithinj replied:
yes, this is a dummy function for now which always returns 1.

subhasisb commented on src/include/mom_func.h:
```diff
@@ -148,6 +148,8 @@ extern enum hup_action	call_hup;
 #ifdef	_PBS_JOB_H
 
 #define COMM_MATURITY_TIME  60 /* time when we consider a pbs_comm connection as mature */
+#define MOM_DELTA_NORMAL	1
```
Please add a comment explaining what these two macros mean

nithinj replied:
done!

subhasisb commented on src/resmom/mom_server.c:
```diff
@@ -753,13 +754,16 @@ is_request(int stream, int version)
 		return;
 	}
 
+	if (server_stream == -1)
```
Is this the case when server asked mom to run a job or delete a job, before the mom had an opportunity to connect to server? Please explain this piece with a block comment?

nithinj replied:
yes, will add the comment.

nithinj replied:
No. they will appear to be down.

subhasisb commented on src/resmom/mom_main.c:
```diff
@@ -9713,7 +9754,7 @@ main(int argc, char *argv[])
 	 * we will send one when we connect to router
 	 */
-	if (pbs_conf.pbs_use_tcp == 0)
+	if (pbs_conf.pbs_use_tcp == 0 && server_stream == -1)
```
We can remove this block since we dont need to send restart to server? Plus pbs_use_tcp is gone if you rebase

subhasisb commented on src/server/mom_info.c:
```diff
@@ -400,6 +400,42 @@ create_svrmom_entry(char *hostname, unsigned int port, unsigned long *pul)
+	DBPRT(("open_momstream"))
+
+	psvrmom = (mom_svrinfo_t *)pmom->mi_data;
+	if (psvrmom->msr_state & INUSE_NEED_ADDRS)
```
What does this mean?

nithinj replied:
Server will multicast all the mom addresses as part of IS_CLUSTERADDR so that inter mom communication can happen securely. So whenever a node is added or deleted, it will mark all nodes down and mark it up once it receives an ack for clusteraddr. If the server is in the middle of that operation, we are not opening a new stream.

I realize that it is anyway handled as we are checking for opened stream in the following line. hence removing.

subhasisb commented on src/server/node_func.c:
```diff
@@ -804,6 +807,12 @@ setup_notification()
 			((mom_svrinfo_t *)(pbsndlist[i]->nd_moms[nmom]->mi_data))->msr_timepinged = 0;
```
Do we still need the msr_timepinged variable at all? We only send IS_NULL if rpp_* values have changed, so that is not a ping

nithinj replied:
nope. removing!

subhasisb commented on src/server/node_manager.c:
```diff
@@ -3245,96 +2953,184 @@ mark_nodes_unknown(int all)
+ */
+void
+mcast_moms(int cmd)
+{
```
So what is the sequence of events when server is already up and we are starting moms in a loop. Each mom will send a HELLOSVR, and server immediately responds with IS_NULL. But then server waits for other moms to join (for 2 seconds) and only after that delay it will send the full sequence of IS_NULL, IS_CLUSTERADDRS etc? Is there no IS_HELLO? On the mom side we still handle IS_HELLO, though (vestige code) ?

nithinj replied:
Mom will immediately reply back with IS_NULL and the exchange will continue.
Mom waits two seconds only before sending clusteraddr as other nodes must be getting created or deleted during the same time. Which make sense as it has to broadcast the message to all the moms. This will not occur in mult-server case.
IS_HELLO is removed. Our upgrade guide suggests the following order:
 shutdown moms -> shutdown server -> start server -> start moms
In that case things should work. But let me know if we need to consider any other upgrade scenario...

subhasisb commented on src/server/node_manager.c:
```diff
@@ -4569,13 +4351,17 @@ is_request(int stream, int version)
 
-		ping_a_mom(pmom, 1, 0);
+		rpp_eom(stream);
+		reply_hellosvr(pmom);
```
Instead of replying to HELLOSVR right away, should we be waiting for the 2 seconds to gang up more moms that might be joining right now and then send a mcast(IS_NULL)? This is what you do later anyway

nithinj replied:
2s waiting is only for clusteraddr message.

I feel an immediate reply is useful for the following:
1. Nodes will be marked up only at the end of the message exchange. Hence better to do it sooner than later.
2. If we broadcast the reply at the end, server will be bombarded with all the response at one shot which is heavier to process and will become unresponsive to other requests. Immediate response will achieve interleaving.
3. mom hello exchange is already sharded and a server has to reply back to only the requests it receives and not to alien moms.

subhasisb commented on src/server/process_request.c:
```diff
@@ -1602,3 +1595,27 @@ get_servername(unsigned int *port)
+	else {
+		if (pbs_conf.pbs_secondary)
+			return parse_servername(pbs_conf.pbs_secondary, port);
+		return NULL;
```
Looks like u are chooing to return NULL (failure) randomly if failover is not configured. Why?

nithinj replied:
caller was checking for that. Refactored this code.

subhasisb commented:
I guess a table review during one of our meetings would be useful

nithinj commented on src/include/mom_func.h:
```diff
@@ -206,6 +210,7 @@ extern char *set_shell(job *, struct passwd *);
 extern void  start_exec(job *);
 extern void  send_obit(job *, int);
 extern void  send_restart(void);
```
done!

nithinj commented on src/include/net_connect.h:
```diff
@@ -86,8 +86,8 @@ typedef unsigned long pbs_net_t;        /* for holding host addresses */
 
-#define	IS_PROTOCOL	4	/* inter-server protocol number */
-#define	IS_PROTOCOL_VER	3	/* inter-server protocol version number */
+#define	IS_PROTOCOL	5	/* inter-server protocol number */
```
This is inter-server communication protocol version. Needed for backward-compatibility. I'm not making use of that in this PR. But helpful if we need to address any compatibility issues due to this change.

nithinj commented on src/lib/Libifl/xml_encode_decode.c:
```diff
@@ -200,7 +200,6 @@ decode_argument(char *encoded_arg, char *original_arg)
 #ifdef WIN32
 	int	quotes_flag = 0;
 #endif
-	DBPRT(("%s: encoded arg= %s\n", __func__, encoded_arg))
```
This was added by some developer for debugging. It will fill up the syslog as it is being called for every client.

nithinj commented on src/lib/Libutil/misc_utils.c:
```diff
@@ -1856,3 +1856,9 @@ delay_query(void)
 }
+
+int 
+get_max_servers()
```
done!

nithinj commented on src/resmom/catch_child.c:
```diff
@@ -1916,139 +1917,92 @@ scan_for_exiting(void)
-	(void)diswui(j, pbs_mom_port);
-	rpp_flush(j);
-	(void)sprintf(log_buffer, "Restart sent to server at %s:%d", svr, port);
+	(void)diswui(stream, pbs_mom_port);
```
done!

nithinj commented on src/resmom/catch_child.c:
```diff
@@ -1916,139 +1917,92 @@ scan_for_exiting(void)
  */
 void
-send_restart(void)
+send_restart()
```
done!

nithinj commented:
@subhasisb , @agrawalravi90 ,
Thanks for the reviews. I've addressed your comments. Please have a look.

agrawalravi90 commented on src/server/mom_info.c:
```diff
@@ -396,6 +395,36 @@ create_svrmom_entry(char *hostname, unsigned int port, unsigned long *pul)
+
+	psvrmom = (mom_svrinfo_t *)pmom->mi_data;
+	if (psvrmom->msr_stream >= 0) {
+		return -1;
```
if it's already open, then I think we should return success instead of failure

nithinj replied:
If the mom stream is open then the mom has marked down by the server/admin due to other reasons. we cannot proceed.

agrawalravi90 commented on src/server/node_func.c:
```diff
@@ -792,6 +791,10 @@ setup_notification()
 	int	nmom;
+	static	time_t last_sent_tm = 0;
+
+	if (get_max_servers() > 1)
```
Can you please add a comment to explain why we are bailing in case there are more than 1 max number of servers?

agrawalravi90 commented on src/server/node_func.c:
```diff
@@ -801,9 +804,14 @@ setup_notification()
+
+	/* send IS_CLUSTERADDR2 to happen in next 2 seconds */
+	if (last_sent_tm < time_now) {
+		last_sent_tm = time_now + 2;
```
Ithink it would be more intuitive to call this 'next_send_tm' as it's the time when it'll be sent next, and check above that time_now > next_send_tm.

agrawalravi90 commented on src/server/node_manager.c:
```diff
@@ -361,11 +359,54 @@ tinsert2(const u_long key1, const u_long key2, mominfo_t *momp, struct tree **ro
+
+	return;
+
+	ret = DIS_NOCOMMIT;
```
control will never reach this line right? you either return before it, or you go to err below it?

agrawalravi90 commented on src/server/node_manager.c:
```diff
@@ -1087,6 +1128,10 @@ send_ip_addrs_to_mom(int stream)
 	if (ret != DIS_SUCCESS)
 		return (ret);
+
+	if (get_max_servers() > 1)
```
Is there any harm doing dis_flush regardless of how many servers there are?

nithinj commented on src/lib/Libutil/misc_utils.c:
```diff
@@ -2158,3 +2158,9 @@ crc_file(char *filepath)
 #endif
 }
+
+int 
```
This is a stub now which will be filled in by the PR which introduces multi-server switches. I'll add the comment header.

nithinj commented on src/resmom/catch_child.c:
```diff
@@ -1938,53 +1939,101 @@ scan_for_exiting(void)
+	if (is_compose(stream, IS_HELLOSVR) != DIS_SUCCESS) {
+		sprintf(log_buffer, "Failed to compose IS_HELLOSVR message");
 		log_err(errno, msg_daemonname, log_buffer);
+		tpp_close(stream);
```
It is being invoked from multiple places and stream is global. handling the error here would be better than propogating back.

nithinj replied:
dis failures are handled in the respective functions rather than propagating as they are widely used on a global stream.

agrawalravi90 replied:
But the caller doesn't know that the stream that it opened has been closed because of an error. I think it's just good coding practice, but if this is how we do it everywhere else then nevermind ..

nithinj commented on src/resmom/mom_main.c:
```diff
@@ -8086,7 +8087,16 @@ net_restore_handler(void *data)
 	if (tpp_log_func)
 		tpp_log_func(LOG_INFO, msg_daemonname, "net restore handler called");
-	send_restart();
+	if (server_stream >= 0) {
```
Nope. Function reuses an open stream, like in the case where server opens the connection first.

nithinj commented on src/resmom/mom_server.c:
```diff
@@ -799,7 +806,14 @@ is_request(int stream, int version)
 			}
-			break;
+
+			if (waiting_hello_resp) {
```
Server can also send when rpp values changes. But it wont send it to the moms where hello exchange is going on as they will seem down to server.

nithinj commented:
@subhasisb & @agrawalravi90 

I've addressed your comments. Please proceed with the review.
You can find the updated design [here](https://pbspro.atlassian.net/wiki/spaces/PD/pages/1457029123/MOM+will+initiate+the+dialogue+sequence+with+server.)

Regression results:

1091 | regression | Tests from given sources on platforms CENTOS7, SUSE15, SUSE42, UBUNTU1804, UBUNTU1604 | nithinj/pbspro@mom_rev  altairengineering/pbspro-extensions@master | 8th Apr 2020 04:39:06.222 PM -04:00 | 00:56:25.88 | 3134 | 0 | 0 | 0 | 0 | 0 | 0 | 3134
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --

agrawalravi90 commented on src/resmom/mom_server.c:
```diff
@@ -285,12 +284,15 @@ free_vnodemap(void)
+registermom(int stream, int combine_msg)
 {
 	int  count = 0;
 	int  hello4_opts = 0;
```
rename to "regmom_opts" or something similar?

agrawalravi90 commented on src/resmom/mom_server.c:
```diff
@@ -316,20 +318,17 @@ reply_hello4(int stream)
-		goto err;
-	ret = diswui(stream, hello4_opts);
-	if (ret != DIS_SUCCESS)
+	if (!combine_msg)
```
you have only 1 call to this function, which passes combine_msg = 1, so this condition will never be true right? Which in this case is dangerous as you'll never do is_compose()

nithinj replied:
yes, keeping the interfaces consistent.

agrawalravi90 commented on src/resmom/mom_server.c:
```diff
@@ -316,20 +318,17 @@ reply_hello4(int stream)
 		goto err;
 
 	/* Send vmap version (when that is ready post 8.0.x */
 	if (hello4_opts & HELLO4_vmap_version) {
```
Rename HELLO4_vmap_version to REGMOM_vmap_version?

agrawalravi90 replied:
There's still 'hello4' in many comments in the function, please rename all of them to registermom

agrawalravi90 commented on src/resmom/mom_server.c:
```diff
@@ -685,6 +683,106 @@ send_hook_checksums(void)
+	int new_retry, new_water;
+	int ret = 0;
+
+	DBPRT(("%s: IS_NULL\n", __func__))
```
you are also calling process_rpp_values for IS_REPLYHELLO, so maybe move this DBPRT() to the switch case for IS_NULL?

nithinj replied:
This function is being invoked from two places. Also, DBPRT is used only by devs.

agrawalravi90 replied:
Ok, I understand what's happening now, IS_REPLYHELLO is sending multiple IS commands together so I guess it still makes sense to print IS_NULL here even when called via IS_REPLYHELLO.

agrawalravi90 commented on src/resmom/mom_server.c:
```diff
@@ -685,6 +683,106 @@ send_hook_checksums(void)
+	u_long	ipdepth = 0;
+	u_long	counter = 0;
+
+	DBPRT(("%s: IS_CLUSTER_ADDRS2\n", __func__))
```
Same comment for this, since you also call this for IS_REPLYHELLO, maybe move the DBPRT to switch case of IS_CLUSTER_ADDRS2

nithinj replied:
This function is being invoked from two places.

agrawalravi90 commented on src/resmom/mom_server.c:
```diff
@@ -765,151 +850,44 @@ is_request(int stream, int version)
-				addrinsert(ipaddr);
-			}
-			if (ret != DIS_EOD)
+			if ((ret = is_compose(stream, IS_REGISTERMOM)) != DIS_SUCCESS)
```
why not just let registermom() do is_compose?

nithinj replied:
We are either composing and flushing in the function itself or delegating both to the caller. This is followed by all the newly introduced functions where there is a need for combining at the caller.

agrawalravi90 commented on src/resmom/mom_server.c:
```diff
@@ -1374,15 +1251,18 @@ hook_requests_to_server(pbs_list_head *plist)
-void
-state_to_server(int what_to_update)
+int
+state_to_server(int what_to_update, int combine_msg)
```
again, it seems like you are always calling this function with combine_msg = 1, do you really need this argument?

nithinj replied:
It is being invoked in multiple places.

nithinj replied:
not really, state_to_server will be invoked with combine_msg = 0 in mom main loop.
The only function which won't be invoked using comine_msg = 0 is registermom. But this is for consistency.
when you invoke with combine_msg = 0:
compose -> encode data -> flush
when you invoke with combine_msg = 1:
just encode data and return

The alternate thing we can do with registermom is compose -> encode data and return
We cannot flush here as we need to combine with another message. Then we will flush in the caller but compose will hide behind a function.

agrawalravi90 commented on src/resmom/mom_server.c:
```diff
@@ -765,151 +850,44 @@ is_request(int stream, int version)
-				}
+			internal_state_update = UPDATE_MOM_STATE;
+			if (need_inv) {
+				if ((ret = state_to_server(UPDATE_VNODES, 1)) != DIS_SUCCESS)
```
don't you need to call is_compose() before calling state_to_server with combine = 1 ?

subhasisb commented:
@nithinj can't we send the data being sent at REGISTERMOM message at HELLOSVR itself? Is there a real need for a separate message at the end?

nithinj commented on src/include/net_connect.h:
```diff
@@ -94,29 +94,21 @@ typedef unsigned long pbs_net_t;        /* for holding host addresses */
  **	Types of Inter Server messages (between Server and Mom).
  */
 #define	IS_NULL			0
-#define	IS_HELLO		1
```
Its no longer referenced anywhere else hence removed.

agrawalravi90 replied:
I know, but since we don't want others to use the missing numbers, I thought that commenting them out instead will help. Maybe add a comment above saying "only use incremental numbers for new IS_ constants, don't use missing ones" ?

nithinj replied:
Server mom exchange can happen only between compatible versions. So it can be re-introduced in a future version as it is not expected to be used against a previous version.

nithinj commented on src/resmom/catch_child.c:
```diff
@@ -2005,35 +1964,69 @@ get_servername_random(unsigned int *port)
+	DBPRT(("Sending hellosvr"))
+
+	if (stream < 0) {
+		svr = get_servername_random(&port);
```
You cannot start mom without a valid server name in conf file. so we should be safe here.

nithinj replied:
I've added a check instead as it can fail if we ever remove the first check which complains when the mom starts today.

nithinj commented on src/resmom/mom_main.c:
```diff
@@ -8087,16 +8087,6 @@ net_restore_handler(void *data)
 	mom_net_up_time = time(0);
 	if (tpp_log_func)
 		tpp_log_func(LOG_INFO, msg_daemonname, "net restore handler called");
-	if (server_stream >= 0) {
```
This is handled in the mom main loop itself. On network failure, the stream will be closed so mom will attempt another hello.

nithinj commented:
@subhasisb ,
As discussed, in this approach where the mom reaching out to server, It is better if the first message is as lighter as it can. As of now, IS_HELLOSVR is just a composed message without anything in its body. It makes sense as it can often get rejected due to node not being created by the admin.
Another alternate would be to send everything as part of hellosvr and process everything in the server. This has the issue that a server might be processing job list from a mom which has been moved to another PBS cluster but not updated the conf file yet.

nithinj commented:
@agrawalravi90 & @subhasisb 
I've addressed your comments. Please continue reviewing.

agrawalravi90 commented on src/server/node_manager.c:
```diff
@@ -3240,88 +2992,137 @@ mark_nodes_unknown(int all)
 
-	if (ptask != NULL)
-		global_ping_task = set_task(WORK_Timed, time_now + ping_nodes_rate, ping_nodes, NULL);
+	if (ret != DIS_SUCCESS) {
```
is this block just for logging? If yes, can you please add a comment for the same? If not can you please add a comment to explain what it's doing?

If it's just for logging, should we only do this at a particular log level? seems like a lot of  computation for just logging.

nithinj replied:
Closing the failed mom streams. will add a comment.

agrawalravi90 commented on src/include/pbs_nodes.h:
```diff
@@ -129,26 +129,25 @@ typedef struct mominfo mominfo_t;
+	int		msr_jbinxsz;  /* size of job index array */
+	struct job	**msr_jobindx;  /* index array of jobs on this Mom */
+	long		msr_vnode_pool;/* the pool of vnodes that belong to this Mom */
+	int		reporting_mom;
```
please add a comment to describe the variable's purpose

agrawalravi90 commented on src/server/node_manager.c:
```diff
@@ -4555,15 +4353,33 @@ is_request(int stream, int version)
 
-		ping_a_mom(pmom, 1, 0);
+		tpp_eom(stream);
+		if (psvrmom->msr_vnode_pool > 0 ? psvrmom->reporting_mom : 1) {
```
this is unnecessarily complex and doesn't really need 2 nested conditional statements, I think you can simply do this:
```suggestion
		if (psvrmom->msr_vnode_pool <= 0 || psvrmom->reporting_mom) {
```

agrawalravi90 commented on src/server/node_manager.c:
```diff
@@ -4555,15 +4353,33 @@ is_request(int stream, int version)
+		if (psvrmom->msr_vnode_pool > 0 ? psvrmom->reporting_mom : 1) {
+			/* mcast reply togethor */
+			add_mom_mcast(pmom, &mtfd_replyhello);
+			static int reply_send_tm = 0;
```
move variable declarations to top of the enclosing block

agrawalravi90 commented on src/server/node_manager.c:
```diff
@@ -4555,15 +4353,33 @@ is_request(int stream, int version)
+			if (reply_send_tm <= time_now) {
+				/* time to wait depends on the no of moms server knows */
+				reply_send_tm = time_now + (mominfo_array_size > 1024 ? 2 : 0);
+				struct work_task *ptask = set_task(WORK_Timed, reply_send_tm, mcast_moms, NULL);
```
same comment here, move it to top of the block

subhasisb commented:
@nithinj please address any pending comments from @agrawalravi90 we go ahead

nithinj commented on src/server/node_manager.c:
```diff
@@ -361,11 +361,132 @@ tinsert2(const u_long key1, const u_long key2, mominfo_t *momp, struct tree **ro
+ * @retval != 0 otherwise.
+ */
+static int
+reply_hellosvr(int stream, int need_inv)
```
this does not have to be combined in the caller.

nithinj commented:
@agrawalravi90 ,
I've addressed the rest of the comments. Please find the response to the PTL comments below:
1) I don't find any benefit in writing test infrastructure around internal messages we are using. These tests will add pain to anybody who wants to change the message exchange in the future. If the communication is not done correctly, even the smoke test will not pass for which we do not need a separate test case.
2) I'd love to write a benchmarking test. But the performance gain can be shown when we have a multi-server scenario and sharding is achieved for initial message exchange, instead of every server reaching out to all the moms. There are improvements for a single server which will be significant if we have a huge number of moms. But writing it up can be done only by checking the node up time. But then as we have moved the hello to mom this time will be delayed. It's hard to devise a test.

Let me know what you think...

agrawalravi90 commented:
> I don't find any benefit in writing test infrastructure around internal messages we are using. These tests will add pain to anybody who wants to change the message exchange in the future. If the communication is not done correctly, even the smoke test will not pass for which we do not need a separate test case.

Does changing the protocol happen often?  I think it's valuable to make sure that the messages are being exchanged in the correct order. When one changes this, they will modify the tests according to the new protocol. You should test various conditions as well, mom going down then coming back up, new mom being added, etc. @subhasisb  what do you think?

About the performance test, I think you should still write one, if it doesn't show much improvement, no worries, it will show better results when we have more of multi-server in. At least it will show that there is no drop in performance and hopefully some benefit at least. Like  starting 500 moms, how long does hello exchange take? Won't that be a useful test?

agrawalravi90 commented:
@nithinj  I don't have any further comments, I just feel the PTL tests are important, but if @subhasisb  disagrees then you can go ahead.

subhasisb commented:
Thanks, merging for now so that others can base their work on top of this. We should get PTL soon, and then we might uncover some more issues, but i guess it is okay for now. Already a good improvement to the code has happened
